pkg/proxy/proxy.go

```diff
@@ -19,28 +19,89 @@ import (
 	"github.com/redhat-appstudio/sprayproxy/pkg/metrics"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/slices"
 )
 
-type BackendsFunc func() []string
+type backend struct {
+	URL string `json:"url"`
+}
 
 type SprayProxy struct {
-	backends    BackendsFunc
-	insecureTLS bool
-	logger      *zap.Logger
+	backends              []string
+	insecureTLS           bool
+	enableDynamicBackends bool
+	logger                *zap.Logger
 }
 
-func NewSprayProxy(insecureTLS bool, logger *zap.Logger, backends ...string) (*SprayProxy, error) {
-	backendFn := func() []string {
-		return backends
-	}
-
+func NewSprayProxy(insecureTLS bool, enableDynamicBackends bool, logger *zap.Logger, backends ...string) (*SprayProxy, error) {
 	return &SprayProxy{
-		backends:    backendFn,
-		insecureTLS: insecureTLS,
-		logger:      logger,
+		backends:              backends,
+		insecureTLS:           insecureTLS,
+		enableDynamicBackends: enableDynamicBackends,
+		logger:                logger,
 	}, nil
 }
 
+func (p *SprayProxy) RegisterBackends(c *gin.Context) {
+	zapCommonFields := []zapcore.Field{
+		zap.String("method", c.Request.Method),
+		zap.String("path", c.Request.URL.Path),
+		zap.String("query", c.Request.URL.RawQuery),
+		zap.Bool("dynamic-backends", p.enableDynamicBackends),
+	}
+	if p.enableDynamicBackends {
```

Comment:
nit - common go style conventions try to reduce the number of indents needed in the main logic blocks. So typically you see `if ! <value>` checks first so that "escape clause" logic appears at the top of the function. I have found this to be more readable.

```suggestion
	if !p.enableDynamicBackends {
	    ...
```

That said, I think we can eliminate this check in the first place (see other comments).